src/pyroboplan/trajectory/trajectory_optimization.py

```diff
@@ -0,0 +1,423 @@
+
+        Parameters
+        ----------
+            q_path : list[array-like]
+                A list of joint configurations describing the desired motion.
+
+        Return
+        ------
+            Optional[pyroboplan.trajectory.polynomial.CubicPolynomialTrajectory]
+                The resulting trajectory, or None if optimization failed
+        """
+        if len(q_path) == 0:
+            warnings.warn("Cannot optimize over an empty path.")
+            return None
+        num_waypoints = self.options.num_waypoints
+        num_dofs = len(q_path[0])
+
+        if len(q_path) == num_waypoints:
+            fully_specified_path = True
+        elif len(q_path) == 2:
+            fully_specified_path = False
+        else:
+            raise ValueError("Path must either be length 2 or equal to num_waypoints.")
+
+        # Preprocess the kinematic limits
+        min_vel = self._process_limits(self.options.min_vel, num_dofs, "min_vel")
+        max_vel = self._process_limits(self.options.max_vel, num_dofs, "max_vel")
+        min_accel = self._process_limits(self.options.min_accel, num_dofs, "min_accel")
+        max_accel = self._process_limits(self.options.max_accel, num_dofs, "max_accel")
+        min_jerk = self._process_limits(self.options.min_jerk, num_dofs, "min_jerk")
+        max_jerk = self._process_limits(self.options.max_jerk, num_dofs, "max_jerk")
+
+        # Initialize the basic program and its variables
+        prog = MathematicalProgram()
+        x = prog.NewContinuousVariables(num_waypoints, num_dofs)
+        x_d = prog.NewContinuousVariables(num_waypoints, num_dofs)
+        xc = prog.NewContinuousVariables(num_waypoints - 1, num_dofs)
+        xc_d = prog.NewContinuousVariables(num_waypoints - 1, num_dofs)
+        h = prog.NewContinuousVariables(num_waypoints - 1)
+
+        # Initial and final conditions
+        if fully_specified_path:
+            for idx in range(num_waypoints):
+                prog.AddBoundingBoxConstraint(q_path[idx], q_path[idx], x[idx, :])
+        else:
+            prog.AddBoundingBoxConstraint(q_path[0], q_path[0], x[0, :])
+            prog.AddBoundingBoxConstraint(q_path[-1], q_path[-1], x[-1, :])
+        # Initial and final velocities should always be zero.
+        prog.AddBoundingBoxConstraint(0.0, 0.0, x_d[0, :])
+        prog.AddBoundingBoxConstraint(0.0, 0.0, x_d[num_waypoints - 1, :])
+
+        for n in range(num_dofs):
+            # Collocation point constraints
+            for k in range(num_waypoints - 1):
+                prog.AddConstraint(
+                    xc[k, n]
+                    == 0.5 * (x[k, n] + x[k + 1, n])
+                    + (h[k] / 8.0) * (x_d[k, n] - x_d[k + 1, n])
+                )
+                prog.AddConstraint(
+                    xc_d[k, n]
+                    == -(1.5 / h[k]) * (x[k, n] - x[k + 1, n])
+                    - 0.25 * (x_d[k, n] + x_d[k + 1, n])
+                )
```

Comment:
Can you add comment for these guys? I'm not sure where they're from.

Reply:
done.